src/libraries/System.Private.CoreLib/src/System/Half.cs

```diff
@@ -0,0 +1,694 @@
+        // Creates Quiet NaN if significand == 0
+        private static Half CreateHalfNaN(bool sign, ulong significand)
+        {
+            const uint NaNBits = ExponentMask | 0x200; // Most significant significand bit
+
+            uint signInt = (sign ? 1U : 0U) << SignShift;
+            uint sigInt = (uint)(significand >> 54);
+
+            return UInt16BitsToHalf((ushort)(signInt | NaNBits | sigInt));
+        }
+
+        private static unsafe Half UInt16BitsToHalf(ushort value)
+            => *(Half*)&value;
+
+        private static ushort RoundPackToHalf(bool sign, short exp, ushort sig)
+        {
+            const int RoundIncrement = 0x8; // Depends on rounding mode but it's always towards closest / ties to even
+            int roundBits = sig & 0xF;
+
+            if ((uint)exp >= 0x1D)
+            {
+                if (exp < 0)
+                {
+                    sig = (ushort)ShiftRightJam(sig, -exp);
+                    exp = 0;
+                }
+                else if (exp > 0x1D || sig + RoundIncrement >= 0x8000) // Overflow
+                {
+                    return sign ? NegativeInfinityBits : PositiveInfinityBits;
+                }
+            }
+
+            sig = (ushort)((sig + RoundIncrement) >> 4);
+            sig &= (ushort)~(((roundBits ^ 8) != 0 ? 0 : 1) & 1);
+
+            if (sig == 0)
+            {
+                exp = 0;
+            }
+
+            return new Half(sign, (ushort)exp, sig)._value;
+        }
+
+        // If any bits are lost by shifting, "jam" them into the LSB.
+        // if dist > bit count, Will be 1 or 0 depending on i
+        // (unlike bitwise operators that masks the lower 5 bits)
+        private static uint ShiftRightJam(uint i, int dist)
+            => dist < 31 ? (i >> dist) | (i << (-dist & 31) != 0 ? 1U : 0U) : (i != 0 ? 1U : 0U);
+
+        private static ulong ShiftRightJam(ulong l, int dist)
+            => dist < 63 ? (l >> dist) | (l << (-dist & 63) != 0 ? 1UL : 0UL) : (l != 0 ? 1UL : 0UL);
+
+        private static ulong DoubleToUInt64Bits(double value)
+            => (ulong)BitConverter.DoubleToInt64Bits(value);
+
+        private static float CreateSingleNaN(bool sign, ulong significand)
+        {
+            const uint NaNBits = float.ExponentMask | 0x400000; // Most significant significand bit
+
+            uint signInt = (sign ? 1U : 0U) << float.SignShift;
+            uint sigInt = (uint)(significand >> 41);
+
+            return CreateSingle(signInt | NaNBits | sigInt);
+        }
+
+        private static double CreateDoubleNaN(bool sign, ulong significand)
+        {
+            const ulong NaNBits = double.ExponentMask | 0x80000_00000000; // Most significant significand bit
+
+            ulong signInt = (sign ? 1UL : 0UL) << double.SignShift;
+            ulong sigInt = significand >> 12;
+
+            return CreateDouble(signInt | NaNBits | sigInt);
+        }
+
+        private static double CreateDouble(ulong value)
```

Comment:
I'm not a fan of having these simple wrappers, but it also makes the code easier to read when compared to `CreateSingle` that takes the individual parts below.

However, I also don't think I have a better suggestion and we shouldn't block the PR on it 😄.

Reply:
Why not just call the BitConverter directly for this one? I think it would make the code even easier to read. When ones see `CreateDouble((ulong)1)`, it is non-obvious that it means CreateDouble from a bit pattern.

Reply:
Cool. I removed the simple `CreateSingle` and `CreateDouble` methods but left the overloads that take in a `bool sign, byte exp, uint sig` in.